Feat: Code Review processes (#8)
opengovsg/engineering-practices: source-control/commits-and-prs.md

history, allowing someone to understand the motivations and nature of
changes in the codebase easily.

## Opening Pull Requests

### PR Description

Pull Requests (PR) are the mechanism through which changes to projects gets reviewed by team members for approval before being incorporated in the codebase.

Writing a good PR description is a critical aspect of software change and historical tracking.

PRs should give all the context a reviewer needs to understand:
- why the PR exists
- the approach taken to solve the issue

The PR description should be generous in the information it provides to help reviewers.
- link to an RFC where other approaches were discussed
- link to compliance policies
- link to issue in github (ideally all PRs relate, either in part or in full, to an issue)
- link to documentation of esoteric API calls being used

Even when reviewers are expected to be from the team, and therefore be familiar with context at that point in time, the PR description should stand on its own from an informational perspective. Doing so ensures:
- The PRs are understandable by other reviewers (e.g. new joiners, other team members)

> - The context will be understandable by your future-self (do not under-estimate the mind's ability to foget details!)

Reviewer: ```suggestion
- The context will be understandable by your future self (do not underestimate the mind's ability to forget details!)
```

Risks typically is a combination of 2 factors:
- **Impact**: what is changed, how many systems or sub systems can thus potentially be affected? Are these system critical? 
  - e.g. "change to the login system": `High`; vs. "change" css font styles on the 'About the team' page: `Low`
- **Likelihood**: does the change concern area with high fluctuations of input or behaviour?

Finally, do follow this guidelines:
  - [Closing issues via Pull Requests](https://help.github.com/en/articles/closing-issues-using-keywords)
  - [Referencing commits, issues and other repos](https://help.github.com/en/articles/autolinked-references-and-urls)
  - [Referencing repository code](https://help.github.com/en/articles/creating-a-permanent-link-to-a-code-snippet)
  - [GitHub Flavored Markdown](https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown)
  - [Marking duplicates](https://help.github.com/en/articles/about-duplicate-issues-and-pull-requests)

### Content and Requirements

> All PRs should include tests. If no tests are present on purpose, the PR description should highlight why none were needed.

Author: ```suggestion
All PRs should include tests. If no tests are present on purpose, the PR description should indicate why none were needed.
```

> Repositories should ideally have explicit code coverage targets, and a policy that no PR reduces coverage. It is recommended that the coverage level be an automatically assessed criteria for a PR to be mergeable (i.e. PR which cause coverage to drop are blocked).

Reviewer: this is good, could we recommend services such as [coveralls.io](https://coveralls.io/github/opengovsg/FormSG?

Author: Sure thing, if we already use it as a standard, might as well write it dow here 👍

Author: Added 🆗

### Draft PRs

> Once a PR is opened, it is expected to be reviewed in a timely manner. Engineers should proactively check the repositories their teams own, and contribute PR reviews before starting entirely new work.

Reviewer: Do you think we can define "timely manner" to one business day? This helps ensure that people are not blocked; it is also a standard inspired by Google and previously communicated to FormSG engineers.

Author: Thanks for the info! If it's already communicated, it should be stated explicitly here. I will amend accordingly.

Author: Amended 👌

Because reviews can take a considerable amount of time, and because engineers are notified when PRs are opened, PR owners should be considerate when opening PRs. PRs should only be opened when the typical criteria have been met:
- PR Description is complete

> - Changeset is complete

Reviewer: Thanks for bringing up the _changeset_, we do encourage the grouping of commits but I think most engineers at OGP have not used this term before.

- Tests are present
- Functional tests have been performed locally

> That being said, it is sometimes needed to get feedback early. For these cases, PRs may be opened in an imcomplete state, provided they meet the following requirements:

Author: ```suggestion
That being said, it is sometimes needed to get feedback early. For these cases, PRs may be opened in an incomplete state, provided they meet the following requirements:
```

1. They are opened as [Draft PRs](https://github.blog/2019-02-14-introducing-draft-pull-requests/)
2. The PR title includes a leading string `[WIP] ` in the PR title
3. [Optional] Add a label `wip`

Draft PRs are not expected to receive proactive reviews from engineers. Instead the PR owner may seek feedback from specific folks as needed.

When the PR is ready for review, the PR owner must transition it to **Ready for Review**, and remove the leading `[WIP]` string from the PR title.

### Closing PRs

All PRs at OGP require (at least) 1 approval to be merged.

> Teams should strive to close PRs in a timely manner. Teams should ideally have an SLA policy for review time and merging. When a repository has many opened PRs, it causes context switching overheads to engineers.

Reviewer: Curious to know - what does such an SLA look like?

Author: I'm not entirely sure to be honest. A threshold could be 2 to 3 weeks (which already seems rather long... but I've seen waaaay worse 😅). Btw, just to clarify, I use the term SLA just to mean there's some value that has been agreed upon "seriously", and so we should specifically have post mortems / retrospectives if it gets breached (or breached too often).

We may not need an SLA for this, if a guideline is sufficient.

The same applies for time-to-first-review we discussed earlier btw, it is up to us to decide whether the 1 business day threshold is a guideline or should have an SLA attached.

Policies for this have a scope beyond the pure management of one PRs btw. They relate to team commitments instead. In a kanban system, a very productive developer may pick many tasks and create many PRs, leading to many opened in parallel. That is not healthy, and not the culture we want however. Instead, features are delivered as a team and ideally there is some upper limit as to how many PRs are opened concurrently. All devs should first enquire with other team members whether they need help before starting something new. In some scrum tools, the task tracking tool itself can remind people of this by having warnings in task swim lanes (e.g. highlight the column red if there are more than X task "in-review" or "in progress").

The policy that apply to a single PR here is within that broader context: if PRs are closed fast, then the chances of high concurrency is low.

Reviewer: >We may not need an SLA for this, if a guideline is sufficient.

I think let's go with encouragement for now

> PR owners are responsible for merging their PR. If a PR is not getting attention from the team, it is up to the PR owner to ask for reviews and approvals. 

Author: ```suggestion
PR owners are responsible for merging their PRs. If a PR is not getting attention from the team, it is up to the PR owner to ask for reviews and approvals. 
```

It is recommended that all repositories adopt a "Stale-PR" policy. PRs which see no activity and are not getting merged should be closed automatically.

If priorities shift, and a PR cannot get the reviews it requires, the PR owner is encouraged to close the PR proactively, and re-open it later, when the team has bandwidth to tend to the work.

## Reviewing Pull Requests

All engineers are required to review PRs as part of their duties. 

### Purpose

PR reviews foster discussions on implementation details, algorithms, feature sets, edge cases, etc. They are an excellent avenue for learning for everyone involved and, can help surface potential issues. Review comments typically touch on subjects such as:
- Feedback on code structure, abstractions, encapsulation, testability, etc.
- Suggest alternative APIs or libraries
- Identify edge cases

> - Gain concensus on structure

Reviewer: Appears to be duplicate of the first point

Author: I've changed the wordings and order a bit, hopefully that works better now

> - Gain concensus on structure

Reviewer: ```suggestion
PR reviews foster discussions on implementation details, algorithms, feature sets, edge cases, etc. They are an excellent avenue for learning for everyone involved and can help surface potential issues. Review comments typically touch on subjects such as:

- Feedback on code structure, abstractions, encapsulation, testability, etc.
- Suggest alternative APIs, libraries, or technologies
- Identify edge cases
- Gain consensus on implementation as a team
```

May I suggest this wording instead? I think it preserves the intent while improving the phrasing slightly

Author: Adopted! 😃

Code is read much more than it is written. PR reviews ensure that changes are legible by multiple people early, before they are integrated into the code base.

> ### Manners

Reviewer: This is a great section. I think we should also urge the introduction of linters, if nitpicks frequently arise.

Reviewer: Ah I see you have already addressed this with the section of code formatters below

PR reviews are done over comment threads in github. As asynchronous discussions, PR comment threads can cause anxiety for both PR owners and reviewers when disagrements surface.

All participants in PR reviews should maintain a level head, and abide by some simple rules of etiquette:
- Assume good intent
- Always be respectful, polite, and pleasant
- Favour questions over statements
- Don't jump to conclusions / criticism
- Be generous in your explanations and background details when proposing alternatives

### Content

Where possible, when a reviewer has an idea for an alternative implementation, he/she should provide some sample code to explain the different approach (use github's [code suggestion](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) feature for that)

> Discussions on code styles preferences are typically not productive in PRs and are discouraged. To make this a non-issue, At OGP, repositories should up opinionated code formatters, which will help to:

Reviewer: Consensus on OGP is to use the [prettier](https://prettier.io/) as the standard opinionated code formatted - we should add this in.

Author: Added 👌

1. Ensure consistent coding style across repos
2. Eliminate discussions on syntax and style in PR review (because that's not useful!)

### WorkFlow

> PR owners are notified of comments as they happen, and this is noisy when a slow trickle of comments is submitted. To prevent that, reviewers should submit all of their comments in one go by using github's review mechanism. This helps reviewers themselves, since a comment or question they had early on may answer itself as they progress through the PR review.

Reviewer: Thanks for this recommendation - I'm guilty of this!

When submitting a review, github classification should be use with care by reviewers, as follows:

> - `Comment`: You have questions or comments and are thus so far non-commital. Should someone else were to approve the PR, you would be generally OK with the PR being merged as-is

Reviewer: ```suggestion
- `Comment`: You have questions or comments and are thus so far non-committal. Should someone else were to approve the PR, you would be generally OK with the PR being merged as-is
```

> - `Approve`: You agree with the changeset in the PR, and stand by it as if you had writtem the code yourself

Author: ```suggestion
- `Approve`: You agree with the changeset in the PR, and stand by it as if you had written the code yourself
```

- `Request changes`:  You have spotted serious flaws in the PR that must be addressed. PR must not be merged in its current state

> The PR owner is expected to answer questions and comments in the PR. When a particular PR comment thread is concluded to the satisfaction of all involved, the thread should be marked resolved through the button `Resolve Conversation`.

Author: Add description of cases where no consensus is reached. How to deal with conflicts?

## Opening Pull Requests

### PR Description

Pull Requests (PR) are the mechanism through which changes to projects gets reviewed by team members for approval before being incorporated in the codebase.

Writing a good PR description is a critical aspect of software change and historical tracking.

PRs should give all the context a reviewer needs to understand:
- why the PR exists
- the approach taken to solve the issue

The PR description should be generous in the information it provides to help reviewers.
- link to an RFC where other approaches were discussed
- link to compliance policies
- link to issue in github (ideally all PRs relate, either in part or in full, to an issue)
- link to documentation of esoteric API calls being used
- link to accademic paper for 

Even when reviewers are expected to be from the team, and therefore be familiar with context at that point in time, the PR description should stand on its own from an informational perspective. Doing so ensures:
- The PR is understandable by other reviewers (e.g. new joiners, other team members)
- The context will be understandable by your future self (do not underestimate the mind's ability to forget details!)

> - The PR description can be used in other processes (e.g. risk assessment for deployments, audits, incident response)

Reviewer 2: This is a great clause to have! If memory serves, this is also somewhat the norm for more established open-source repos

Author: Yes, using PRs to other processes is quite common. It can be a reason to adopt a SOP for merging btw, since ideally, you want to be able to retrieve the PRs from the repo commits easily.

Just for reference, at Zendesk we always use merge commits, never squash, and we opted to use github's default merge messages, which are of this form:
```
Merge pull request #12345 from <branch_name>
```

That allowed several of our tools to easily find the PRs from commits, and then extract information from the PR description (in particular extracting the risk section). But yes, it does mean this tools were working on convention, e.g.:
* regexp for merge commits messages: `/^Merge pull request #(\d+)/`
* description would have risk at bottom, with a title `##Risks` and everything below could be extracted as the description fo the risks

> To faciliate writing consistent PRs descriptions, all repositories should have an up to date [PR template](https://docs.github.com/es/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository). PRs should have the following sections:

Reviewer 2: nit(grammar):
```suggestion
To faciliate writing consistent PR descriptions, all repositories should have an up to date [PR template](https://docs.github.com/es/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository). PRs should have the following sections:
```

- Problem
- Solution
- Tests
- Risks

> The Risk section is one of the most important part of the PR. The PR author indicates therein he/she has done the required thinking of what could go wrong (best case - worst case) when the code is live, how errors might cascade through the system, and how likely it really is to happen. Risk is a subjective assessment, and obviously, no one opens PRs thinking things are likely to break. Still, all changes carry risk, and engineers are notoriously bad at evaluating risks. The risk section is a reminder for the PR author to think about risks. It also helps reviewers validate the thinking, and comment with their own evaluation if necessary.

Reviewer 2: nit(grammar):
```suggestion
The Risk section is one of the most important parts of the PR. The PR author indicates therein he/she has done the required thinking of what could go wrong (best case - worst case) when the code is live, how errors might cascade through the system, and how likely it really is to happen. Risk is a subjective assessment, and obviously, no one opens PRs thinking things are likely to break. Still, all changes carry risk, and engineers are notoriously bad at evaluating risks. The risk section is a reminder for the PR author to think about risks. It also helps reviewers validate the thinking, and comment with their own evaluation if necessary.
```

- Assume good intent
- Always be respectful, polite, and pleasant
- Favour questions over statements
- Don't jump to conclusions / criticism
- Be generous in your explanations and background details when proposing alternatives

### Content

Where possible, when a reviewer has an suggestion for an alternative implementation, he/she should provide some sample code to explain the different approach (use github's [code suggestion](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) feature for that)

Discussions in PRs on code style are typically not productive and are discouraged. To make this a non-issue, repositories should adopt opinionated code formatters, which will help to ensure a consistent coding style across repos, and eliminate discussions on syntax and style. At OGP, [prettier](https://prettier.io/) has been adopted as the organisation-wide formatter. All repos should use it, and teams are strongly encouraged to set up formatter pre-commit hooks with [husky](https://typicode.github.io/husky/).

> ### WorkFlow

Reviewer 2: nit(casing):
```suggestion
### Workflow
```

- Explain in different wordings or give alternative examples or code samples that illustrate your points better
- Indicate you understand the alternative proposal or system, but you don't agree with it

As asynchronous written discussions, comment threads do not convey tone of voice or body language. When disagreement occurs, It can be immensely helpful to call for a face to face meeting, in person or over zoom, to go over the contentious item live. An agreement can more easily be reached in person.

Disagreement sometimes occur when both parties understand each-other solutions well, but disagree on the tradeoffs that should apply for a particular situation, and that is what driving their divide on the implementation. Such cases should probably be discussed in the broader team, or escalated. The PM or Tech Lead may have to clarify which trade off we wish to pursue as a team.

When a solution is eventually picked, you may still disagree with it, but your commitment to deliver on the decision is expected anyway.

## Merging Pull Requests

PRs at OGP require at least 1 approval in order to be merged.

> In case of contributions from open source contributors, or contributions from egineers not members of the team owning the repo, one approval from a member of the owning team should be gathered.

Reviewer 2: ```suggestion
In case of contributions from open source contributors, or contributions from engineers who are not members of the team owning the repo, one approval from a member of the owning team should be gathered.
```

- Stay polite and respectful at all time
- Do not reply when flustered
- Consider the position of the other developer carefully to see the merits/flaws of their arguments
- Explain in different wordings or give alternative examples or code samples that illustrate your points better
- Indicate you understand the alternative proposal or system, but you don't agree with it

As asynchronous written discussions, comment threads do not convey tone of voice or body language. When disagreements occur, it can be immensely helpful to call for a face to face meeting, in person or over zoom, to go over the contentious item live. An agreement can more easily be reached in person.

Disagreement sometimes occur when both parties understand each-other solutions well, but disagree on the tradeoffs that should apply for a particular situation, and that is what driving their divide on the implementation. Such cases should probably be discussed in the broader team, or escalated. The PM or Tech Lead may have to clarify which trade off we wish to pursue as a team.

When a solution is eventually picked, you may still disagree with it, but your commitment to deliver on the decision is expected anyway.

> A note on the [law of triviality](https://en.wikipedia.org/wiki/Law_of_triviality) (a.k.a [bike-shedding](https://bikeshed.org/)): it has been noted that "the amount of noise generated by a change is inversely proportional to the complexity of the change". Simple changes are easy to understand and a reviewer may quickly form an opinion on how he/she would have done it (slightly?) differently. Resist the urge to propose trivial changes that do not fundamentally affect the good working, reliability, or security of the PR. If you do feel like giving a comment anyway, mark it as "nit" comment to indicate it is not a blocker to acquire your approval to the PR. And if the PR author rejects your change, do not take it personally, and do not block the PR. This is unfortunately a grey area. Some seeminlgy trivial proposals could indeed make the code "better" (say, more readable for example), but it hard to tell definitively when something is subjective vs. objective. Just being aware of what bike-shedding is will make your reviews better.

Reviewer 2: ```suggestion
A note on the [law of triviality](https://en.wikipedia.org/wiki/Law_of_triviality) (a.k.a [bike-shedding](https://bikeshed.org/)): it has been noted that "the amount of noise generated by a change is inversely proportional to the complexity of the change". Simple changes are easy to understand and a reviewer may quickly form an opinion on how he/she would have done it (slightly?) differently. Resist the urge to propose trivial changes that do not fundamentally affect the good working, reliability, or security of the PR. If you do feel like giving a comment anyway, mark it as "nit" comment to indicate it is not a blocker to acquire your approval to the PR. And if the PR author rejects your change, do not take it personally, and do not block the PR. This is unfortunately a grey area. Some seemingly trivial proposals could indeed make the code "better" (say, more readable for example), but it hard to tell definitively when something is subjective vs. objective. Just being aware of what bike-shedding is will make your reviews better.
```

The PR description should be generous in the information it provides to help reviewers.
- link to an RFC where other approaches were discussed
- link to compliance policies
- link to issue in github (ideally all PRs relate, either in part or in full, to an issue)
- link to documentation of esoteric API calls being used
- link to accademic paper for 

Even when reviewers are expected to be from the team, and therefore be familiar with context at that point in time, the PR description should stand on its own from an informational perspective. Doing so ensures:
- The PR is understandable by other reviewers (e.g. new joiners, other team members)
- The context will be understandable by your future self (do not underestimate the mind's ability to forget details!)
- The PR description can be used in other processes (e.g. risk assessment for deployments, audits, incident response)

To faciliate writing consistent PR descriptions, all repositories should have an up to date [PR template](https://docs.github.com/es/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository). PRs should have the following sections:

> - Problem
> - Solution

Author: ```suggestion
- Context
- Approach
```

@liangyuanruo @LoneRifle : Any objection if I change the 2 headers per the suggestion above? I think that is more generic, where PRs can deliver improvements or new features, without spawning from a "problem". 

If no objection, I'll make a PR accordingly in ts-template, I need to make a PR to add the risk section there anyway.
